src/walberla_bridge/src/lattice_boltzmann/LBWalberlaImpl.hpp

```diff
@@ -1753,14 +1753,14 @@ class LBWalberlaImpl : public LBWalberlaBase {
       std::vector<int> const &raster_flat) const override {
     Utils::Vector3d force({0, 0, 0});
     auto const &grid_size = get_lattice().get_grid_dimensions();
-    for (auto const &block : *get_lattice().get_blocks()) {
+    for (auto &block : *get_lattice().get_blocks()) {
       auto const offset = get_lattice().get_block_corner(block, true);
       auto const &force_field = m_boundary->get_force_vector(&block);
       auto const &index_field = m_boundary->get_index_vector(&block);
       for (int i = 0; i < raster_flat.size(); i++) {
         if (raster_flat[i] != 0) {
           auto node = flat_index_to_node(i);
-          if (get_lattice().node_in_local_domain(node)) {
+          if (get_lattice().node_in_local_halo(node)) {
```

Comment:
I had to change from local domain to local halo, otherwise contributions would be lost when a boundary stretches across a periodic boundary. To reproduce the error with the original code, run `lb_buoyancy_force.py` with sphere parameters `RADIUS = 0.8 * AGRID` and `center=self.system.box_l / 2 + 0.5 * AGRID`. This will create a single boundary cell on the periodic boundary (`print(self.lbf[8:11,8:11,8:11].is_boundary.astype(int))`). Integrate for a few time steps and print the force on the boundary, using 1 or 2 MPI ranks to see the difference in the per-shape force.